File: app/src/main/java/net/sytes/kai_soft/letsbuyka/CRUDdb.java
package net.sytes.kai_soft.letsbuyka;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import net.sytes.kai_soft.letsbuyka.CustomList.CustomList;
import net.sytes.kai_soft.letsbuyka.Lists.List;
import net.sytes.kai_soft.letsbuyka.ProductModel.Product;

/**
 * Created by Лунтя on 06.06.2018.
 */
/*  !!!ЗДЕСЬ НАДО ДОБАВИТЬ ЧТЕНИЕ ЭЛЕМЕНТОВ БД. СЕЙЧАС ОНИ В АДАПТЕРАХ!!!*/

/*  Класс в котором можно производить добавление, считывание, обновление и удаление из таблиц
 *   базы данных */
public class CRUDdb {
    private static DataBase db = Application.Companion.getDB();   //Экземпляр базы данных

    /*  Добавление продукта в таблицу продуктов. Принимает имя и описание продукта  */
    public static void insertToTableProducts(String name, String description) {
        ContentValues cv = new ContentValues();

        //db = Application.getDB();  // Хз зачем я это сделал

        SQLiteDatabase db = CRUDdb.db.getWritableDatabase();

        cv.put(DataBase.tableProducts.TABLE_ITEM_NAME, name);
        cv.put(DataBase.tableProducts.TABLE_DESCRIPTION, description);

        db.insert(DataBase.TABLE_NAME_PRODUCTS_LIST, null, cv);
    }

    /*  Обновление элемента списка в таблице продуктов  */
    public static void updateTableProducts(Product product) {
        ContentValues cv = new ContentValues();

        //db = Application.getDB(); // Зачем я опять это сделал?!

        SQLiteDatabase db = CRUDdb.db.getWritableDatabase();

        cv.put(DataBase.tableProducts.TABLE_ITEM_NAME, product.getItemName());
        cv.put(DataBase.tableProducts.TABLE_DESCRIPTION, product.getDescription());
        //cv.put(DataBase.tableProducts.TABLE_PHOTO, product.getFirstImagePath());

        db.update(DataBase.TABLE_NAME_PRODUCTS_LIST, cv,
                DataBase.tableProducts.TABLE_ID + " = ?",
                new String[]{String.valueOf(product.getId())});
    }

    /*  Удаление элемента списка из таблицы продуктов   */
    public static void deleteItemProducts(Product product) {
        SQLiteDatabase db = CRUDdb.db.getWritableDatabase();

        db.delete(DataBase.TABLE_NAME_PRODUCTS_LIST,
                DataBase.tableProducts.TABLE_ID + " = ?",
                new String[]{String.valueOf(product.getId())});
    }


    /*  Добавление списка в таблицу списков. Принимает имя списка  */
    public static void insertToTableLists(String name) {
        ContentValues cv = new ContentValues();

        //db = Application.getDB(); //И ещё раз хз.

        SQLiteDatabase db = CRUDdb.db.getWritableDatabase();

        cv.put(DataBase.tableLists.TABLE_ITEM_NAME, name);

        db.insert(DataBase.TABLE_NAME_LISTS_LIST, null, cv);
    }

    /*  Обновление списка в таблице списков    */
    public static void updateTableLists(List list) {
        ContentValues cv = new ContentValues();

        //db = Application.getDB(); //Чёт плохо как-то...

        SQLiteDatabase db = CRUDdb.db.getWritableDatabase();

        cv.put(DataBase.tableLists.TABLE_ITEM_NAME, list.getItemName());

        db.update(DataBase.TABLE_NAME_LISTS_LIST, cv,
                DataBase.tableLists.TABLE_ID + " = ?",
                new String[]{String.valueOf(list.getId())});
    }

    /*  Удаление списка из таблицы списков */
    public static void deleteItemLists(List list) {
        SQLiteDatabase db = CRUDdb.db.getWritableDatabase();

        db.delete(DataBase.TABLE_NAME_LISTS_LIST, DataBase.tableLists.TABLE_ID + " = ?",
                new String[]{String.valueOf(list.getId())});
    }


    /*  Добавление элемента в таблицу пользовательских списков  */
    public static void insertToTableCustomList(long id_list, long id_product) {
        ContentValues cv = new ContentValues();

        //db = Application.getDB(); //...

        SQLiteDatabase SQLdb = CRUDdb.db.getWritableDatabase();

        /*  Здесь идёт проверка на наличие такой записи в этом пользовательском списке.
         *   Если такая запись есть, то её добавлять не нужно.*/
        Cursor c = SQLdb.rawQuery("select * from " +
                        DataBase.TABLE_NAME_CUSTOM_LIST + " where " +
                        DataBase.tableCustomList.TABLE_ID_LIST + " = " + String.valueOf(id_list) +
                        " and " + DataBase.tableCustomList.TABLE_ID_PRODUCT + " = " +
                        String.valueOf(id_product),
                null);
        if (c.moveToFirst() == false) {
            cv.put(DataBase.tableCustomList.TABLE_ID_LIST, id_list);
            cv.put(DataBase.tableCustomList.TABLE_ID_PRODUCT, id_product);
            cv.put(DataBase.tableCustomList.TABLE_DEPRECATED,
                    CustomList.DEPRECATED_FALSE);
            // вставляем запись и получаем ее ID
            SQLdb.insert(DataBase.TABLE_NAME_CUSTOM_LIST, null, cv);
        }
        c.close();
    }

    /*  Удаление элемента из таблицы пользовательских списков   */
    public static void deleteItemCustomList(long id) {
        SQLiteDatabase db = CRUDdb.db.getWritableDatabase();

        db.delete(DataBase.TABLE_NAME_CUSTOM_LIST, DataBase.tableLists.TABLE_ID + " = ?",
                new String[]{String.valueOf(id)});
    }

    /*  Зачёркиваем элемент в пользовательском списке   */
    public static void makeDeprecated(long id) {

        ContentValues cv = new ContentValues();

        //db = Application.getDB(); //Почему, мистер Андерсон, почему?

        SQLiteDatabase db = CRUDdb.db.getWritableDatabase();

        Cursor c = db.rawQuery("select * from " + DataBase.TABLE_NAME_CUSTOM_LIST + " where " +
                DataBase.tableCustomList.TABLE_ID + " = " + String.valueOf(id), null);

        int idDeprecatedColumn = c.getColumnIndex(DataBase.tableCustomList.TABLE_DEPRECATED);
        if (c.moveToFirst()) {
            /*Если элемент не был зачёркнут, зачёркиваем. Или наоборот*/
            if (c.getLong(idDeprecatedColumn) == CustomList.DEPRECATED_TRUE) {
                cv.put(DataBase.tableCustomList.TABLE_DEPRECATED,
                        CustomList.DEPRECATED_FALSE);
            } else {
                cv.put(DataBase.tableCustomList.TABLE_DEPRECATED,
                        CustomList.DEPRECATED_TRUE);
            }

            db.update(DataBase.TABLE_NAME_CUSTOM_LIST, cv,
                    DataBase.tableCustomList.TABLE_ID + " = ?",
                    new String[]{String.valueOf(id)});
        }
        c.close();
    }

}

File: app/src/main/java/net/sytes/kai_soft/letsbuyka/CRUDdb.kt
package net.sytes.kai_soft.letsbuyka

import android.annotation.SuppressLint
import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.util.Log
import net.sytes.kai_soft.letsbuyka.Lists.List

import net.sytes.kai_soft.letsbuyka.ProductModel.Product

@Suppress("DUPLICATE_LABEL_IN_WHEN")
/*  CRUD class for DataBase*/
class CRUDdb {
    companion object {
        private val dataBase: DataBase = Application.getDB()
        private const val tag = "letsbuy_CRUDdb"

        /*  Add product */
        fun insertToTableProducts(name: String, description: String) {
            Log.i(tag, "insertToTableProducts($name, $description)")

            val cv = ContentValues()

            val db = CRUDdb.dataBase.writableDatabase

            cv.put(Constants.TABLE_ITEM_NAME, name)
            cv.put(Constants.TABLE_DESCRIPTION, description)

            db.insert(Constants.TABLE_NAME_PRODUCTS, null, cv)
        }

        /*  Update product  */
        fun updateTableProducts(product: Product) {
            Log.i(tag, "updateTableProducts($product)")

            val cv = ContentValues()

            val db = CRUDdb.dataBase.writableDatabase

            cv.put(Constants.TABLE_ITEM_NAME, product.name)
            cv.put(Constants.TABLE_DESCRIPTION, product.description)

            db.update(Constants.TABLE_NAME_PRODUCTS, cv,
                    Constants.TABLE_ID + " = ?",
                    arrayOf(product.id.toString()))
        }

        /*  Delete product  */
        fun deleteItemProducts(product: Product) {
            Log.i(tag, "deleteItemProducts($product)")

            val db = CRUDdb.dataBase.writableDatabase

            db.delete(Constants.TABLE_NAME_PRODUCTS,
                    Constants.TABLE_ID + " = ?",
                    arrayOf(product.id.toString()))
        }

        /*  Initial filling products. Works only first start    */
        fun fillProducts(context: Context, dataBase: SQLiteDatabase) {
            Log.i(tag, "fillProducts()")

            val cv = ContentValues()

            val product = context.resources.getStringArray(R.array.productsList)

            for (name: String in product) {
                cv.put(Constants.TABLE_ITEM_NAME, name)
                cv.put(Constants.TABLE_DESCRIPTION, "")
                dataBase.insert(Constants.TABLE_NAME_PRODUCTS, null, cv)
            }
        }


        /*  Add list    */
        fun insertToTableLists(name: String) {
            Log.i(tag, "insertToTableLists($name)")

            val cv = ContentValues()

            val db = CRUDdb.dataBase.writableDatabase

            cv.put(Constants.TABLE_ITEM_NAME, name)

            db.insert(Constants.TABLE_NAME_LISTS, null, cv)
        }

        /*  Read from "lists" with filter   */
        @SuppressLint("Recycle")
        fun readFromTableLists(filter: String? = null): ArrayList<List> {
            Log.i(tag, "readFromTableLists($filter)")

            val dataBase = CRUDdb.dataBase.readableDatabase

            val lists = ArrayList<List>()
            val c = if (filter == null) {
                dataBase.query(Constants.TABLE_NAME_LISTS, null, null, null, null, null,
                        Constants.TABLE_ID + " asc")
            } else {
                dataBase.rawQuery("select * from " + Constants.TABLE_NAME_LISTS +
                        " where " + Constants.TABLE_ITEM_NAME + " like '%"
                        + filter + "%'",
                        null)

            }

            if (c.moveToFirst()) {
                val idColIndex = c.getColumnIndex(Constants.TABLE_ID)
                val nameColIndex = c.getColumnIndex(Constants.TABLE_ITEM_NAME)
                do {
                    lists.add(List(c.getInt(idColIndex), c.getString(nameColIndex)))
                } while (c.moveToNext())
            }
            c.close()
            return lists
        }

        /*  Update list */
        fun updateTableLists(list: List) {
            Log.i(tag, "updateTableLists($list)")

            val cv = ContentValues()

            val dataBase = CRUDdb.dataBase.writableDatabase

            cv.put(Constants.TABLE_ITEM_NAME, list.itemName)

            dataBase.update(Constants.TABLE_NAME_LISTS, cv,
                    Constants.TABLE_ID + " = ?",
                    arrayOf(list.id.toString()))
        }

        /*  Delete list */
        fun deleteItemLists(list: List) {
            Log.i(tag, "deleteItemLists($list)")

            val dataBase = CRUDdb.dataBase.writableDatabase

            dataBase.delete(Constants.TABLE_NAME_LISTS, Constants.TABLE_ID + " = ?",
                    arrayOf(list.id.toString()))
        }


        /*  Add element to custom list  */
        fun insertToTableCustomList(idList: Int, idProduct: Int) {
            Log.i(tag, "insertToTableCustomList($idList,$idProduct)")

            val dataBase = CRUDdb.dataBase.writableDatabase

            val cv = ContentValues()

            val c = dataBase.rawQuery("select * from " +
                    Constants.TABLE_NAME_CUSTOM_LIST + " where " +
                    Constants.TABLE_ID_LIST + " = " + idList.toString() +
                    " and " + Constants.TABLE_ID_PRODUCT + " = " +
                    idProduct.toString(),
                    null)
            if (!c.moveToFirst()) {
                cv.put(Constants.TABLE_ID_LIST, idList)
                cv.put(Constants.TABLE_ID_PRODUCT, idProduct)
                cv.put(Constants.TABLE_DEPRECATED, Constants.DEPRECATED_FALSE)

                dataBase.insert(Constants.TABLE_NAME_CUSTOM_LIST, null, cv)
            }
            c.close()
        }

        /*  Delete element from custom list */
        fun deleteItemCustomList(id: Int) {
            Log.i(tag, "deleteItemCustomList($id)")

            val dataBase = CRUDdb.dataBase.writableDatabase

            dataBase.delete(Constants.TABLE_NAME_CUSTOM_LIST,
                    Constants.TABLE_ID + " = ?", arrayOf(id.toString()))

        }

        /*  Crossing out element*/
        fun makeDeprecated(id: Int) {
            Log.i(tag, "makeDeprecated($id)")

            val dataBase = CRUDdb.dataBase.writableDatabase
            val cv = ContentValues()
            val c = dataBase.rawQuery("select * from " + Constants.TABLE_NAME_CUSTOM_LIST
                    + " where " + Constants.TABLE_ID + " = " + id.toString(), null)
            val idDeprecatedColumn = c.getColumnIndex(Constants.TABLE_DEPRECATED)
            if (c.moveToFirst()) {
                if (c.getInt(idDeprecatedColumn) == Constants.DEPRECATED_TRUE) {
                    cv.put(Constants.TABLE_DEPRECATED, Constants.DEPRECATED_FALSE)
                } else {
                    cv.put(Constants.TABLE_DEPRECATED, Constants.DEPRECATED_TRUE)
                }
                dataBase.update(Constants.TABLE_NAME_CUSTOM_LIST, cv,
                        Constants.TABLE_ID + " = ?", arrayOf(id.toString()))
            }
            c.close()
        }

    }
}